### PR TITLE
fix(sql): fix async filter with cursor returning bad results

### PR DESCRIPTION
**Pull request by bziobrowski**
This PR fixes handling of subquery where clauses on repeated filter compilation for async filter factory . 
In master branch first compilation (main filter instance) can alter (even  remove) nested model's where clause so next (per-worker) compilations will produce different results.
Query result's correctness depends on whether original- or per-worker-  filter is used.

Example : 

```sql
CREATE TABLE price (
  ts TIMESTAMP,
  type SYMBOL,
  value DOUBLE 
) TIMESTAMP(ts) PARTITION BY DAY;

CREATE TABLE mapping ( 
  id SYMBOL, 
  ext SYMBOL, 
  ext_in SYMBOL, 
  ts timestamp 
) TIMESTAMP(ts);

INSERT INTO price 
SELECT x::timestamp,  't' || (x%5), rnd_double()  
FROM long_sequence(100000);

INSERT INTO mapping 
SELECT 't' || x, 's' || x, 's' || x, x::timestamp  
FROM long_sequence(5);

SELECT count(*) 
FROM price 
WHERE type IN  (SELECT id FROM mapping WHERE ext in ('s1', 's2') and substring(ext,2,1) = '1');
```

Sql above should return 20000 but can return arbitrary number on current master branch.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 21 / 21 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|21|21|100.00%|

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java [hunk 1]
```diff
@@ -409,6 +418,7 @@ private Function compileFilter(
         if (condition) {
             ObjList<Function> workerFilters = new ObjList<>();
             for (int i = 0; i < workerCount; i++) {
+                restoreWhereClause(filterExpr);//restore original filters in node query models
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java:421]: nit: `restoreWhereClause` should use start index `1` for accessing join models, `model.joinModels.get(0) == model` all the time

Comment (bluestreak01) [reply core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java:421]: fixed in https://github.com/questdb/questdb/pull/2958/commits/fdd798b752780c214589698d1cab4711e67af2f1

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 23 / 23 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/model/QueryModel.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|21|21|100.00%|

**Pull request merged**